src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -206,44 +207,99 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
             var itemsWithPatternMatch = items.SelectAsArray(
                 item => (item, completionHelper.GetMatch(item.FilterText, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture)));
 
-            return FilterItems(completionHelper, itemsWithPatternMatch);
+            return FilterItems(completionHelper, itemsWithPatternMatch, filterText);
         }
 
         internal static ImmutableArray<CompletionItem> FilterItems(
             CompletionHelper completionHelper,
-            ImmutableArray<(CompletionItem item, PatternMatch? match)> itemsWithPatternMatch)
+            ImmutableArray<(CompletionItem item, PatternMatch? match)> itemsWithPatternMatch,
+            string filterText)
         {
-            var bestItems = ArrayBuilder<(CompletionItem, PatternMatch?)>.GetInstance();
-            foreach (var pair in itemsWithPatternMatch)
+            // It's very common for people to type expecting completion to fix up their casing,
+            // so if only lowercase characters were typed so far, we'd loosen our standard on comparing items
+            // in case-sensitive manner and take into consideration the MatchPriority as well.
+            // i.e. when everything else is equal, then if item1 is a better case-sensitive match but item2 has higher 
+            // MatchPriority, we consider them equally good match, so the controller will later have a chance to
+            // decide which is the best one to select.
+            var isFilterTextLowercaseOnly = !filterText.Any(char.IsUpper);
+
+            // Keep track the highest MatchPriority of all items in the best list.
+            var highestMatchPriorityInBest = int.MinValue;
+            using var _1 = ArrayBuilder<(CompletionItem, PatternMatch?)>.GetInstance(out var bestItems);
+
+            // This contains a list of items of worse case-sensitive match but higer MatchPriority than all best ones.
+            using var _2 = ArrayBuilder<(CompletionItem, PatternMatch?)>.GetInstance(out var similarItemsWithHigerMatchPriority);
+
+            foreach (var (item, match) in itemsWithPatternMatch)
```

Comment:
i foudn this all a bit confusing.  can you doc the contract of the FilterItems method?  specifically, i can see we get a bunch of items/matches, and we have the filtertext.  What is the contract around the array of items we get back?   Having that contract will help me understand what the logic here is supposed to be doing and how we can ensure we have hte right behavior with good perf.  Thanks! 

I'm skipping this mehtod for now.

Reply:
It seems we are expecting `FilterItems` to determine among the items the best match w.r.t. the given filter text, and those returned would be considered equally good candidates for selection.

Then our ItemManager would either pick the one in MRU if there's any, or the one with highest MatchPriority.